src/merkle_tree.rs

```diff
@@ -0,0 +1,1383 @@
+// This file contains the code used to create a full MerkleSet and is heavily reliant on the code in merkle_set.rs.
+
+use clvmr::sha2::{Digest, Sha256};
+use crate::merkle_set::{
+    hash,
+    BLANK,
+    get_bit,
+    NodeType,
+};
+#[cfg(test)]
+use crate::merkle_set::compute_merkle_set_root;
+#[cfg(test)]
+use rand::rngs::SmallRng;  // cargo says this isn't required but tests won't run without it
+#[cfg(test)]
+use rand::{Rng, SeedableRng};
+#[cfg(feature = "py-bindings")]
+use pyo3::{pyclass, pymethods, PyResult};
+#[cfg(feature = "py-bindings")]
+use pyo3::exceptions;
+#[cfg(feature = "py-bindings")]
+use pyo3::types::{PyList, PyBytes};
+#[cfg(feature = "py-bindings")]
+use pyo3::prelude::*;  // needed for PyBytes
+
+// the ArrayType is used to create a more lasting MerkleTree representation in the MerkleSet struct
+#[derive(PartialEq, Debug, Copy, Clone)]
+pub enum ArrayTypes {
+    Leaf { data: usize },                // indexes for the data_hash array
+    Middle { children: (usize, usize) }, // indexes for a Vec of ArrayTypes
+    Empty,
+    Truncated,
+}
+
+// represents a MerkleTree by putting all the nodes in a vec. Root is the last entry.
+#[derive(PartialEq, Debug, Clone)]
+#[cfg_attr(
+    feature = "py-bindings",
+    pyclass(frozen, name = "MerkleSet"),
+)]
+pub struct MerkleSet {
+    pub nodes_vec: Vec<ArrayTypes>,
+    pub leaf_vec: Vec<[u8; 32]>,
+    pub hash_cache: Vec<[u8; 32]>, // same size and order as nodes_vec
+}
+
+const EMPTY: u8 = 0;
+const TERMINAL: u8 = 1;
+const TRUNCATED: u8 = 2;
+const MIDDLE: u8 = 3;
+
+const EMPTY_NODE_HASH: [u8; 32] = [127, 156, 158, 49, 172, 130, 86, 202, 47, 37, 133, 131, 223, 38, 45, 188, 125, 111, 104, 242, 160, 48, 67, 213, 201, 154, 74, 229, 167, 57, 108, 233];
+
+
+#[derive(Debug)]
+#[cfg_attr(
+    feature = "py-bindings",
+    pyclass(frozen, name = "SetError"),
+)]
+pub struct SetError;
+
+pub fn deserialize_proof(proof: &[u8]) -> Result<MerkleSet, SetError> {
+    let mut merkle_tree: MerkleSet = MerkleSet::default();
+    let pos = _deserialize(&mut merkle_tree, proof, 0)?;
+    if pos != proof.len() {
+        Err(SetError)
+    } else {
+        Ok(merkle_tree)
+    }
+}
+
+fn _deserialize(
```

Comment:
Since things in Rust are private by default, this leading underscore is not necessary. It typically indicates that something is unused, rather than internal.